### PR TITLE
add abstractions for commands that require files

### LINKED ISSUE #71: Create UiParser, move help/exit into UiParser
UiParser should handle functions that involve Ui functionality, rather than going through AddressBookParser.

Also Export/Import failures must be shown in the commandBox.

### PR DESCRIPTION
**Pull request by guangxuan**
Adds abstractions for functionality of commands that require files:

- `CommandPreAction` holds information required by `MainWindow` to perform `UiOperation` before command execution, such as asking for a file.
- `CommandWithFile` represents commands that require a file to execute. Not covariant to Command unfortunately.
- `parseCommandWithFile` added to `AddressBookParser` to handle parsing of `CommandWithFile`.
- `ExportBuyersCommand`, `ExportPropertiesCommand`, `ExportCommandParser` added to follow the existing workflow.
- `ExportPropertiesCommand` execute calls `Storage` which calls `CsvManager`. Returns a `CommandResult` like other commands.
- `CommandResult` modified to hold UiAction as an enum to indicate the type of action Ui has to do after execution of command.

closes #71 and #78

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/ExportCommand.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package seedu.address.logic.commands;
+
+public abstract class ExportCommand extends CommandWithFile {
+    public static final String COMMAND_WORD = "export";
+
+    public static final String PROPERTIES = "properties";
+
+    public static final String BUYERS = "buyers";
```
Comment (sunjc826) [inline src/main/java/seedu/address/logic/commands/ExportCommand.java:8]: Should there be a central source of the words "property", "buyer", "properties", "buyers"?

File: src/main/java/seedu/address/logic/Logic.java [hunk 2]
```diff
@@ -18,22 +18,36 @@
...
      */
     CommandResult execute(String commandText) throws CommandException, ParseException;
 
+
     /**
-     * Exports properties to csv file.
+     * Executes the command and returns the result.
+     *
+     * @param commandText The command as entered by the user.
+     * @param file        File chosen by the user.
```
Comment (sunjc826) [inline src/main/java/seedu/address/logic/Logic.java:34]: Is this argument description alignment consistent with the rest of the comments?
What do the others think?

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 3]
```diff
@@ -62,15 +69,56 @@ public Command parseCommand(String userInput) throws ParseException {
...
+        }
+        final String commandWord = matcher.group("commandWord");
+        final String objectType = matcher.group("objectType");
+        switch (commandWord) {
+        case ExportCommand.COMMAND_WORD:
+            switch (objectType) {
+            case ExportCommand.PROPERTIES:
+                return Optional.of(new ExportPropertiesCommand());
+            case ExportCommand.BUYERS:
+                return Optional.of(new ExportBuyersCommand());
```
Comment (sunjc826) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:103]: Is this consistent with how parsing is done by the regular parseCommand? The `objectType` in the regular parseCommand is retrieved from the preamble in ParserUtil.

File: src/main/java/seedu/address/logic/Logic.java (hunk 2 above)
Comment (waynetee) [reply src/main/java/seedu/address/logic/Logic.java:34]: can refer to this https://github.com/nus-cs2103-AY2122S1/forum/issues/10

File: src/main/java/seedu/address/logic/commands/ExportCommand.java (hunk 1 above)
Comment (guangxuan) [reply src/main/java/seedu/address/logic/commands/ExportCommand.java:8]: Should have some sources here and there. It's ok to repeat words. 
Will consider moving Properties and Buyers in the next PR to a more general Command.

File: src/main/java/seedu/address/logic/Logic.java (hunk 2 above)
Comment (guangxuan) [reply src/main/java/seedu/address/logic/Logic.java]: Reverted to old style.

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java (hunk 3 above)
Comment (guangxuan) [reply src/main/java/seedu/address/logic/parser/AddressBookParser.java]: Created an ExportComnandParser

File: src/main/java/seedu/address/logic/LogicManager.java [hunk 4]
```diff
@@ -57,21 +60,22 @@ public CommandResult execute(String commandText) throws CommandException, ParseE
...
-        } catch (IOException ioe) {
-            logger.warning("Problem while exporting Properties.");
+    public CommandResult execute(String commandText, File file) throws CommandException, ParseException {
+        logger.info("----------------[USER COMMAND][" + commandText + "]");
+
+        CommandResult commandResult;
+        Optional<CommandWithFile> command = addressBookParser.parseCommandWithFile(commandText);
+        if (command.isEmpty()) {
+            throw new CommandException("execute commandWithFile is empty!");
         }
```
Comment (waynetee) [inline src/main/java/seedu/address/logic/LogicManager.java:70]:
```suggestion
        assert !command.isEmpty() : "commandText is invalid";
```

File: src/main/java/seedu/address/logic/commands/ExportCommand.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Exports buyers or properties from csv file.\n"
+            + "Examples: " + COMMAND_WORD + " " + BUYERS + "\n"
+            + COMMAND_WORD + " " + PROPERTIES + "\n";
+
+    public static final String MESSAGE_CANCELLED = "Export %s was cancelled.";
+
+    public static final String MESSAGE_FAILURE = "Export %s failed to save the file.";
```
Comment (waynetee) [inline src/main/java/seedu/address/logic/commands/ExportCommand.java:16]:
```suggestion
    public static final String MESSAGE_FAILURE = "Export %s failed to save the file. Try closing other apps that might have this file opened.";
```

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 6]
```diff
@@ -26,6 +30,8 @@
      * Used for initial separation of command word and args.
      */
     private static final Pattern BASIC_COMMAND_FORMAT = Pattern.compile("(?<commandWord>\\S+)(?<arguments>.*)");
+    private static final Pattern COMMAND_WITH_FILE_FORMAT =
+            Pattern.compile("(?<commandWord>\\S+) +(?<objectType>\\S+)(?<arguments>.*)");
```
Comment (waynetee) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:34]:
```suggestion
```

**Review by waynetee — APPROVED** [review]
LGTM with minor suggestions, good job!
Thanks for the patience with the multiple rounds of refactoring too
👍🏻

File: src/main/java/seedu/address/logic/Logic.java [hunk 7]
```diff
@@ -18,22 +19,32 @@
 public interface Logic {
     /**
      * Executes the command and returns the result.
+     *
      * @param commandText The command as entered by the user.
      * @return the result of the command execution.
      * @throws CommandException If an error occurs during command execution.
-     * @throws ParseException If an error occurs during parsing.
+     * @throws ParseException   If an error occurs during parsing.
```
Comment (sunjc826) [inline src/main/java/seedu/address/logic/Logic.java:26]: Do take note that there is still a space here.

File: src/main/java/seedu/address/logic/Logic.java [hunk 8]
```diff
@@ -18,22 +19,32 @@
...
+
     /**
-     * Exports properties to csv file.
+     * Executes the command and returns the result.
+     *
+     * @param commandText The command as entered by the user.
+     * @param file File chosen by the user.
+     * @return the result of the command execution.
+     * @throws CommandException If an error occurs during command execution.
+     * @throws ParseException   If an error occurs during parsing.
```
Comment (sunjc826) [inline src/main/java/seedu/address/logic/Logic.java:38]: As well as for these 2 lines.

**Pull request merged**